### PR TITLE
Podboat: Support file deletion

### PR DESCRIPTION
**Pull request by dennisschagt**
Implements feature request from issue #669.

### CODE REVIEW

File: src/queueloader.cpp [hunk 1]
```diff
@@ -214,6 +215,20 @@ void QueueLoader::reload(std::vector<Download>& downloads, bool remove_unplayed)
...
+	for (const auto& dl : deletion_list) {
+		std::string filename = dl.filename();
+		if (access(filename.c_str(), F_OK) == 0) {
+			LOG(Level::INFO,
+				"Deleting file %s",
+				filename);
+			if (std::remove(filename.c_str()) != 0) {
+				LOG(Level::ERROR,
+					"Failed to delete file %s",
+					filename);
```
Comment (Minoru) [inline src/queueloader.cpp]: It'd be useful to log `errno` here, ideally both as an integer and as text.

File: src/queueloader.cpp [hunk 2]
```diff
@@ -206,6 +215,22 @@ void QueueLoader::reload(std::vector<Download>& downloads, bool remove_unplayed)
 		f.close();
 	}
 
+	if (ctrl->get_cfgcont()->get_configvalue_as_bool("delete-downloaded-files")) {
+		for (const auto& dl : deletion_list) {
+			std::string filename = dl.filename();
```
Comment (Minoru) [inline src/queueloader.cpp]: Please use `const` if the object is not modified.

File: src/queueloader.cpp [hunk 3]
```diff
@@ -206,6 +215,22 @@ void QueueLoader::reload(std::vector<Download>& downloads, bool remove_unplayed)
...
+			if (access(filename.c_str(), F_OK) == 0) {
```
Comment (Minoru) [inline src/queueloader.cpp]: This creates a [TOCTOU bug](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use), but luckily this check can be avoided: we can just run `std::remove`, check its exit code, and inspect `errno`, logging errors only if `errno != ENOENT`.

File: src/queueloader.cpp (hunk 1 above)
Comment (dennisschagt) [reply src/queueloader.cpp]: Good idea, done

File: src/queueloader.cpp (hunk 3 above)
Comment (dennisschagt) [reply src/queueloader.cpp]: Oops, fixed

**Comment by dennisschagt** [comment]
> I'm doubtful about the new setting's name, I need to sleep on it.

I don't like the setting's name either so I'm happy if you come up with something better. (I didn't think of a better name yet)

> Also, this feature is a good candidate to be unit-tested. We don't have any coverage for QueueLoader yet, but FileUrlReader is quite similar, and you can take a look at this test as an example. It's fine if you don't—I only require tests for modules that already have some tests, so this PR is fine without.

I'm planning to write some tests but before that I think some small refactoring is useful to make writing tests easier.
I think it is better if I open a separate pull request for writing test+refactor.

**Comment by Minoru** [comment]
I think the setting's name should be `delete-played-files`. It's a bit misleading, since finished files are deleted as well, but that can be amended with docs.

> I'm planning to write some tests but before that I think some small refactoring is useful to make writing tests easier.

That's risky: you might inadvertently break something during refactoring, and won't know about it because there are no tests. Writing tests first doesn't eliminate the risk completely, but lowers it (significantly IMHO).

You can try your approach, but before submitting the PR, please run a simple check: remove all refactorings (with interactive rebase, or by cherry-picking tests onto "master"), and make sure the tests pass.

It might be useful to look at code coverage while writing tests. You can either submit a PR marked as draft, and view Coveralls reports online (the bot will post a link after CI build); or run `test/generate_coverage_report.sh` locally.

> I think it is better if I open a separate pull request for writing test+refactor.

Yeah, definitely. Thanks!

**Comment by dennisschagt** [comment]
Done

**Comment by Minoru** [comment]
Thank you, @dennisschagt!

**Pull request merged**